### ISSUE TITLE
SideBar should list, not have hard coded transclusions.

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#2739]
In core, I find the practice to display content by *listing based on tag*, and then transclude the content, to be very good. However, the [sidebar](http://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FPageTemplate%2Fsidebar) (link leads to tw.com sidebar) instead *hard codes* the transclusions. This makes it difficult to manipulate the sidebar display or one must overwrite it. 

**Proposal**: The sidebar should list its content based on tags.

---

**Elaboration and example**

The sidebar is *hard coded* to transclude...

* site title
* site subtitle
* page controls
* sidebar lists

I (often!) wish to decorate my site titles, for instance with some image but occasionally with more. For limited cases, I can hook on to the site title via css trickery. Another way is to, via the controlpanel, manually type in the site title and a transclusion or link. This is clumsy. Instead, I should simply tag a tiddler with something and use *list-after* or *list-before* do decide its position in the sidebar.

**Actual case**: I'm just fiddling with a decorated "sign" stating *"Beta"* that should appear on all my beta stage releases and I want to release this as a plugin. It should also have a link so people can click it to find more info. But it can't be a plugin currently because there is no way to hook on to the site title to show it there automatically.

**Comment by danielo515** [comment]
Totally agree . This is one of the few non modular parts of the fore

**Comment by Jermolene** [comment]
Fixing this is just a matter of agreeing on some conventions, and then rearranging some wikitext core templates; it would therefore be a good case for a pull request from people who don't normally consider themselves core contributors...

The conventions we need include things like tag such as `$:/tags/SideBarSection` to identify the sections of the sidebar (`$:/tags/SideBar` is already taken; it should be `$:/tags/SideBarTab`)

The complexities include thing like catering for people who want to customise the sidebar back to the old pre-popup behaviour.

**Comment by danielo515** [comment]
>it would therefore be a good case for a pull request from people who don't normally consider themselves core contributors...

Should I feel quoted ? :smile:

**Comment by twMat** [comment]
I say we find some middle-aged, bald guy with a beard and a Scandinavian accent for this job. In case we don't find such a guy, I'll do it.

>The complexities include thing like catering for people who want to customise the sidebar back to the old pre-popup behaviour.

@Jermolene I don't understand the practical consequences of this. Which popups? The search results? And how do we cater for those who "want to customize" - they can just customize by overwriting it, no?

**Comment by twMat** [comment]
BTW, in the sidebar tiddler there are *classes* appointed to the various tiddlers.  in the sidebar elements, can they be moved to inside their respective tiddlers?

**Comment by tobibeer** [comment]
If possible, I would appreciate if we do not use a tagging mechanism, at least not until a tiddler body can be upgraded independent from any tagging.

**Comment by twMat** [comment]
@tobibeer - so what do you suggest as alternative?

**Comment by tobibeer** [comment]
Good question. Maybe just some list field or even the body of some config tiddler, e.g. `$:/config/components/Sidebar`. I mean, why would tagging have to play a role in this? Simply list the components you wish to be rendered in the order you want them to... in one spot. Simple.

**Comment by twMat** [comment]
>why would tagging have to play a role in this?

I partly agree but that is the practice we have for these things - or do we use any other method for comparable cases somewhere?

One idea with the OP is that users can control the sidebar content, without overwriting core stuff. If the list is hard coded, this breaks it... unless it is coded to *also* aggregate user-added tiddlers... which, btw, would also have to be identified somehow...

Have you anywhere elaborated on:
>at least not until a tiddler body can be upgraded independent from any tagging.

It makes sense, but I have another angle on it so I wouldn't mind discussing it.

**Comment by Jermolene** [comment]
> If possible, I would appreciate if we do not use a tagging mechanism, [...]

@tobibeer I understand that you have reservations about using tags for configuring the UI, but it makes no sense to change things in mid-stream for this one case, leaving all of the other system tags working as before. If we did as you propose we would be adding to the entropy of TiddlyWiki, introducing a new mechanism that isn't used consistently for the same purpose elsewhere.

TW5 currently uses system tags to compose UI components. We will continue to use the same technique unless and until we decide to change things across the board.

(I wish somebody would start experimenting with completely replacing the page template, and building their own TW UI from scratch; that would be a perfect opportunity to experiment with other techniques).

> [...] at least not until a tiddler body can be upgraded independent from any tagging.

I don't understand what you mean there.

> I mean, why would tagging have to play a role in this? Simply list the components you wish to be rendered in the order you want them to... in one spot. Simple.

Using tagging for the UI was an integral part of the original TW5 design; it was a reaction against the MainMenu of TWClassic. The goal was to enable multiple plugins to add UI components independently. That wouldn't be possible if we just listed the components in one place.

**Comment by tobibeer** [comment]
> [...] at least not until a tiddler body can be upgraded independent from any tagging.
> I don't understand what you mean there

I meant that right now, just by removing a system tag a tiddler becomes part of the wiki rather than be a shadow... and so upgrading it isn't easily possible while preserving the state where `$:/tags/ViewTemplate` would be toggled off, at the moment of upgrading.

> Using tagging for the UI was an integral part of the original TW5 design; it was a reaction against the MainMenu of TWClassic. The goal was to enable multiple plugins to add UI components independently. That wouldn't be possible if we just listed the components in one place.

I simply wanted to raise the thought that with list-before and list-after as well as the list field, we've added so much more complexity to configuring layout that a simple list seems could have helped avoid.

I fully understand that changing these paradigmns now is a hard task or...as you say... requires a different layouting paradigm.

**Comment by pmario** [comment]
<I simply wanted to raise the thought that with list-before and list-after as well as the list field, we've added so much more complexity to configuring layout that a simple list seems could have helped avoid.

tobias is right here

**Comment by Jermolene** [comment]
> tobias is right here

@tobibeer @pmario actually I disagree. Although the tag mechanism is complicated when you take ordering into account, it answers specific needs around merging customisations from multiple sources. I'm open to improvements, but not to regressions that make things worse. Falling back to a simple list would be making things much, much worse.

Also, I disagree with the way that this has been raised in relation to this specific issue. Surely nobody seriously thinks that we should unilaterally use a different technique for some random part of the UI than the standards we have already established? If it's just a matter of raising a discussion point then it would be better in a separate issue. Lets keep this issue for @twMat's original proposal of making the sidebar programmable.

**Comment by Jermolene** [comment]
Actually to be clearer: I find that it is mostly a waste of time for me personally to gett too involved in "doom and gloom" discussions about some existing core mechanism to be broken beyond clarifying the current behaviour (because the criticism so frequently misses the point of the original implementation, as here). However it's important that I do pay attention to actual user issues, such as the one that @twMat has raised here. That's why I'd much prefer that simple, actionable issues like this should not be hijacked for philosophical discussions.

**Comment by danielo515** [comment]
I have to agree on this with @Jermolene 

1. Let's keep this thread about the original proposal
1. We should not start using a different mechanism for this proposal. This arises from the necessity make TW more consistently configurable, not to experiment with new ways of generating layouts

@Jermolene could you please expand on the required tags that you envisioned for this feature? as long as I read they are `$:/tags/SideBarTab` for adding new tabs to the sidebar and `$:/tags/SideBarSection` to add new sections to the entire sidebar, right?

If someone removes the tag `$:/tags/SideBarSection` from the tabbed interface, and then it adds `$:/tags/SideBarTab` this can cause confusion to the user because he can not see the tiddler as a sidebar tab. I'm not saying that we should take care of this scenario, I'm just advancing one possible scenario of a confused user.

Regards

**Comment by tobibeer** [comment]
> Surely nobody seriously thinks that we should unilaterally use a different technique for some random part of the UI than the standards we have already established?

I agree. Actually, using just the list field does precisely do what I was proposing... the thing that's missing is a more simple ui, ideally drag-sortable for maintaining list items in the list field, or any list-like field for that matter. If all is sorted there, no list-before or after is needed. However, it leaves us with the "problem" that tagging triggers layout, rather than "listing"... so, there's no easy way to say: consider only those titles in the list field of **$:/tags/ViewTemplate** for layouting. Not sure, will have to think about whether we can easily build a switch that caters for the latter.

**Comment by twMat** [comment]
@Jermolene 

Ok, so I made [this](http://newsidebar.tiddlyspot.com/)

It works - but for some reason I cannot get the list sequence right!? I've tried with various `list` setups but with only partial success.

The password is set to `password`

**Comment by Jermolene** [comment]
Hi @twMat that's a great start.

I do think that the site title and subtitle should be separate tiddlers.

You've put the `list` field on `$:/core/ui/PageTemplate/sidebar`; it should actually go on the tag tiddler `$:/tags/SideBarSection` (which you'll need to add).

You shouldn't need `list-*` fields on the individual sections.

**Comment by twMat** [comment]
>I do think that the site title and subtitle should be separate tiddlers.
>
>You've put the list field on $:/core/ui/PageTemplate/sidebar; it should actually go on the tag tiddler $:/tags/SideBarSection (which you'll need to add).
>
>You shouldn't need list-* fields on the individual sections.

Ok, [fixed](http://newsidebar.tiddlyspot.com/).  (Thanx for note about where to put the list field!)

I also fixed some misses that affected the styling.

**Comment by Jermolene** [comment]
Hi @twMat looking good.

Now we need to make it into a pull request, and include the documentation :smile:

Are you ready to give that a go? I can't remember where you got to on your last Node.js experiments.

Otherwise, I can put together a pull request for you from the TiddlySpot wiki.

**Comment by twMat** [comment]
I can attempt to write documentation for it. For a PR, I'd appreciate it if I don't - at this time - have to get involved with Node.js. So, please @Jermolene or anyone, feel welcome to take it.

**Comment by Jermolene** [comment]
@danielo515 @tobibeer @Marxsal @pmario - is anyone able to volunteer to help @twMat turn his TiddlySpot wiki into a pull request?

**Comment by Marxsal** [comment]
I've made a PR for the basic changes -- not sure if I did it all right so check closely. If I'm supposed to be doing plugin stuff, I'll need some pointers how that works with github.

Maybe someone who understands the consequences better can update docs ??

**Comment by twMat** [comment]
@Marxal - big thank you :-)

The idea is not for it to be a plugin so I'm guessing what you've done is basically good. 

To not mess things up, I'm stating the following bits instead of trying to modify your PR:

No idea if this matters: Your "tid files" have a capital initial on the last "word", e.g "core/ui/SideBar/**L**ists.tid" but the actual tiddlers have lower case there.

Apropos titles; I should have titled `$:/core/ui/SideBar/pagecontrols` to instead be `$:/core/ui/SideBar/pagetools` (ref. "the Page Toolbar"). 

Again, thank you @Marxsal

**Comment by Marxsal** [comment]
I think the case must not be too important, since I see it both ways in the existing TW5 file set. I made them title case in order to match the existing tids in the Sidebar directory. Maybe Jeremy can wade in on his preferences. I can change "pagecontrols" then as well since we would have the same case question.

Thanks!

**Comment by Jermolene** [comment]
Thanks @telmiger
